core/src/main/java/ai/timefold/solver/core/impl/solver/termination/AdaptiveTermination.java

```diff
@@ -0,0 +1,171 @@
+                @NonNull Score_ end) {
+            var scoreDiffs = end.subtract(start).toLevelDoubles();
+            var softestLevel = scoreDiffs.length - 1;
+            for (int i = 0; i < scoreDiffs.length; i++) {
+                if (scoreDiffs[i] != 0.0) {
+                    return new LevelScoreDiff(softestLevel != i, scoreDiffs[softestLevel]);
+                }
+            }
+            return null;
+        }
+    }
+
+    public void start(long startTime, Score_ startingScore) {
+        resetGracePeriod(startTime, startingScore);
+    }
+
+    public void step(long time, Score_ bestScore) {
+        scoresByTime.put(time, bestScore);
+    }
+
+    private void resetGracePeriod(long currentTime, Score_ startingScore) {
+        gracePeriodStartTimeMillis = currentTime;
+        isGracePeriodActive = true;
+
+        // Remove all entries in the map since grace is reset
+        scoresByTime.clear();
+
+        // Put the current best score as the first entry
+        scoresByTime.put(currentTime, startingScore);
+    }
+
+    public boolean isTerminated(long currentTime, Score_ endScore) {
+        if (isGracePeriodActive) {
+            // first score in scoresByTime = first score in grace period window
+            var endpointDiff = LevelScoreDiff.between(scoresByTime.firstEntry().getValue(), endScore);
+            var timeElapsedMillis = currentTime - gracePeriodStartTimeMillis;
+            if (endpointDiff != null && endpointDiff.harderScoreChanged) {
+                // A harder score changed, so reset the grace period
+                resetGracePeriod(currentTime, endScore);
+                return false;
+            }
+            if (timeElapsedMillis >= gracePeriodMillis) {
+                // grace period over, record the reference diff
+                isGracePeriodActive = false;
+                gracePeriodScoreDiff = endpointDiff;
+                return endpointDiff == null;
+            }
+            return false;
+        }
+
+        var startScoreEntry = scoresByTime.floorEntry(currentTime - gracePeriodMillis);
+        var startScore = startScoreEntry.getValue();
+        scoresByTime.subMap(0L, startScoreEntry.getKey()).clear();
+        var scoreDiff = LevelScoreDiff.between(startScore, endScore);
+
+        if (scoreDiff == null) {
+            // no change after grace period, terminate
+            return true;
+        }
+
+        if (scoreDiff.harderScoreChanged) {
+            // A harder score changed, so reset the grace period
+            resetGracePeriod(currentTime, endScore);
+            return false;
+        }
+
+        return scoreDiff.softestScoreDiff / gracePeriodScoreDiff.softestScoreDiff < minimumImprovementRatio;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public boolean isSolverTerminated(SolverScope<Solution_> solverScope) {
+        return isTerminated(System.currentTimeMillis(), (Score_) solverScope.getBestScore());
```

Comment:
I wonder if we should use nanotime instead, and truncate it to millis.
The precision of `currentTimeMillis` is questionable, see Javadoc:

> Note that while the unit of time of the return value is a millisecond, the granularity of the value depends on the underlying operating system and may be larger. For example, many operating systems measure time in units of tens of milliseconds.

Reply:
Also... should this terminate the solver even in, for example, a CH phase? Arguably it should not, the CH phase generally doesn't improve.

The way how unimproved terminations get around it is by forcing the adaptive termination to be only possible in local search phase. Maybe we do something like that too?

Although I really don't like that, because it causes config pain. (You need to define LS phase, and put the termination inside of it.) Maybe we refactor the terminations a bit, so that it is possible for an unimproved termination to be defined on the solver-level, but never terminate a CH? Would need a discussion, I think.

(My ultimate goal is to have only one termination. This one. And this configuration quirk is preventing that from happening.)

Reply:
It cannot terminate in the CH phase, because the CH ALWAYS improve (namely, it improves the init score every step, which resets the grace period).

Reply:
Not necessarily true - list var with allows unassigned has init score always zero.
But doesn't matter - this termination, for now, is only allowed on the LS phase.

Reply:
CH in unassigned does not decrease the best score; it only can increase it (it can leave the value unassigned (no increase), or assign a value (which more than likely will increase the medium score).

Reply:
Those are assumptions - as long as the thing is technically possible, we need to have an answer. In this case, this (along with the unimproved terminations) will be exempt from terminating the CH, once it's even possible to configure them on the top level. For now, until we implement that, it's not a problem at all.

In fact, I am surprised why this termination even cares about init score. How does it receive events from CH, if it needs to be configured on the LS only? That smells like a bug.

Reply:
It was from `solvingStarted` (where the best score is typically uninitialized). All terminations gets the `solvingStarted` event even if configured on a phase level.

Reply:
(in short, the termination was incorrectly using `solvingStarted` instead of `phaseStarted` to initialize).